File: Java/Sorting/src/com/company/CycleSort.java
// for numbers from 1 to (n-1)
package com.company;

import java.util.Arrays;

public class CycleSort {
    public static void main(String[] args) {
        int[] arr = {5, 4, 3, 2, 1};
        cycleSort(arr);
        System.out.println(Arrays.toString(arr));
    }
    static void cycleSort(int[] arr) {
        int i = 0;
        while (i < arr.length) {
            int correctIndex = arr[i] - 1;
            if(arr[i] != arr[correctIndex]) {
                swap(arr, i, correctIndex);
            }else {
                i++;
            }
        }
    }
    static void swap(int[] arr, int first, int second) {
        int temp = arr[first];
        arr[first] = arr[second];
        arr[second] = temp;
    }
}


// Time complexity
// Best case = O(n)
// Worst case = O(n)

File: C++/Sorting/Cycle_sort.cpp
// Cycle sort
// Time complexity = O(N)
// For numbers for 1 to (n - 1)

#include <iostream>
using namespace std;

void swapped(int arr[], int first, int second)
{
    int temp = arr[first];
    arr[first] = arr[second];
    arr[second] = temp;
}

void cycleSort(int arr[], int size)
{
    int i = 0;
    while (i < size)
    {
        int correctindex = arr[i] - 1;
        if (arr[i] != arr[correctindex])
        {
            swapped(arr, i, correctindex);
        }
        else
        {
            i++;
        }
    }
}

int main()
{
    int n;
    cout << "Enter the number of elements in the array: " << endl;
    cin >> n;
    int arr[n], ans[n];
    cout << "Enter the array: " << endl;
    for (int i = 0; i < n; i++)
    {
        cin >> arr[i];
    }
    cout << "The array is: " << endl;
    for (int j = 0; j < n; j++)
    {
        cout << arr[j] << " ";
    }
    cout << endl;
    cycleSort(arr, n);
    cout << "The sorted array is: " << endl;
    for (int k = 0; k < n; k++)
    {
        cout << arr[k] << " ";
    }
    cout << endl;
    return 0;
}